tutorials/multi/1_graph_classification.py

```diff
@@ -0,0 +1,241 @@
+
+def init_model(seed, device):
+    torch.manual_seed(seed)
+    model = GIN().to(device)
+    model = DistributedDataParallel(model, device_ids=[device], output_device=device)
+
+    return model
+
+###############################################################################
+# Main Function for One Process
+# -----------------------------
+#
+# Define the model evaluation function as in the single-GPU setting.
+#
+
+def evaluate(model, dataloader, device):
+    model.eval()
+
+    total = 0
+    total_correct = 0
+
+    for bg, labels in dataloader:
+        bg = bg.to(device)
+        labels = labels.to(device)
+        # Get input node features
+        feats = bg.ndata.pop('attr')
+        with torch.no_grad():
+            pred = model(bg, feats)
+        _, pred = torch.max(pred, 1)
+        total += len(labels)
+        total_correct += (pred == labels).sum().cpu().item()
+
+    return 1.0 * total_correct / total
+
+###############################################################################
+# Define the main function for one process. Since different processes handle
+# different data batches, some of them can finish their task earlier.
+# `dist.barrier()` forces faster processes to wait until all processes
+# have encountered the barrier.
+#
+
+from torch.optim import Adam
+
+def main(rank, world_size, dataset, seed=0):
+    init_process_group(world_size, rank)
+    # Assume the GPU ID to be the same as the process ID
+    device = torch.device('cuda:{:d}'.format(rank))
+    torch.cuda.set_device(device)
+
+    model = init_model(seed, device)
+    criterion = nn.CrossEntropyLoss()
+    optimizer = Adam(model.parameters(), lr=0.01)
+
+    train_loader, val_loader, test_loader = get_dataloaders(dataset,
+                                                            seed,
+                                                            world_size,
+                                                            rank)
+    for epoch in range(5):
+        model.train()
+        # The line below ensures all processes use a different
+        # random ordering in data loading for each epoch.
+        train_loader.set_epoch(epoch)
+
+        total_loss = 0
+        for bg, labels in train_loader:
+            bg = bg.to(device)
+            labels = labels.to(device)
+            feats = bg.ndata.pop('attr')
+            pred = model(bg, feats)
+
+            loss = criterion(pred, labels)
+            total_loss += loss.cpu().item()
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+        loss = total_loss
+        print('Loss: {:.4f}'.format(loss))
+        dist.barrier()
+
+        val_acc = evaluate(model, val_loader, device)
+        print('Val acc: {:.4f}'.format(val_acc))
+        dist.barrier()
+
+    test_acc = evaluate(model, test_loader, device)
+    print('Test acc: {:.4f}'.format(test_acc))
+    dist.barrier()
+    dist.destroy_process_group()
```

Comment:
Why do we destroy in `main` instead of `'__main__'`?

Reply:
Honestly I'm not sure and simply following the practice in the PyTorch tutorial [here](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html). This might be related to that `init_process_group` is called per process.

Reply:
OK that's fair.